File: Gdx/Historic/fell/core/src/com/dugsolutions/fell/map/gen2/GenZMound.java
package com.dugsolutions.fell.map.gen;

import com.badlogic.gdx.math.Vector3;

/**
 * A center point where the height is at max.
 * 
 * Then a slope defines a parabolic slope outward toward the edge, max circle
 * radius, where the value is zero.
 */
public class GenZMound extends GenZCone {
	protected float mPA;

	public GenZMound(float height) {
		super(height);
	}
	
	public GenZMound(int count, float height) {
		super(count, height);
	}

	@Override
	protected void setZ(int x, int y) {
		float dX = x - fCenterX;
		float dY = y - fCenterY;
		float dXSquared = dX * dX;
		float dYSquared = dY * dY;
		float height;
		float dist = (float) Math.sqrt(dXSquared + dYSquared);

		if (mIsCircle) {
			height = -mPA * dist * dist + mHeight;
		} else {
			float maxDist;

			if (dX != 0) {
				float angleT = (float) Math.atan(dY / dX);

				if (dX < 0) {
					angleT += Math.PI;
				}
				maxDist = (float) getDistOnEllipse(angleT);
			} else {
				maxDist = fB;
			}
			// Parabola along the line in the ellipsoid of interest.
			height = -(mHeight / (maxDist * maxDist)) * dist * dist + mHeight;
		}
		if (height > 0) {
			icalc.addHeight(x, y, height);

			if (icalc.genNormal()) {
				Vector3 normal;
				normal = new Vector3(dX, dY, -height);
				normal.nor();
				icalc.addNormal(x, y, normal);
			}
		}
	}

	@Override
	protected void init(int startx, int starty, int endx, int endy) {
		super.init(startx, starty, endx, endy);
		mPA = mHeight / (fA * fA);
	}

}

File: Gdx/FellDungeon/app/src/main/java/com/dugsolutions/fell/map/gen2/GenZMound.kt
package com.dugsolutions.fell.map.gen2

import com.badlogic.gdx.math.Vector3
import kotlin.math.atan

/**
 * A center point where the height is at max.
 *
 * Then a slope defines a parabolic slope outward toward the edge, max circle
 * radius, where the value is zero.
 */
open class GenZMound : GenZCone {
    protected var mPA = 0f

    constructor(height: Float) : super(height)
    constructor(count: Int, height: Float) : super(count, height)

    override fun setZ(x: Int, y: Int) {
        val dX = x - fCenterX
        val dY = y - fCenterY
        val dXSquared = dX * dX
        val dYSquared = dY * dY
        val height: Float
        val dist = Math.sqrt((dXSquared + dYSquared).toDouble()).toFloat()
        if (mIsCircle) {
            height = -mPA * dist * dist + super.height
        } else {
            val maxDist: Float
            if (dX != 0f) {
                var angleT = atan((dY / dX).toDouble()).toFloat()
                if (dX < 0) {
                    angleT += Math.PI.toFloat()
                }
                maxDist = getDistOnEllipse(angleT.toDouble()).toFloat()
            } else {
                maxDist = fB
            }
            // Parabola along the line in the ellipsoid of interest.
            height = -(super.height / (maxDist * maxDist)) * dist * dist
        }
        if (height > 0) {
            val ic = icalc ?: return
            ic.addHeight(x, y, height)
            if (ic.genNormal()) {
                val normal = Vector3(dX, dY, -height)
                normal.nor()
                ic.addNormal(x, y, normal)
            }
        }
    }

    override fun init(startx: Int, starty: Int, endx: Int, endy: Int) {
        super.init(startx, starty, endx, endy)
        mPA = height / (fA * fA)
    }
}